### PR TITLE
[pickers] Fix `DigitalClock` time options on a `DST` switch day

### LINKED ISSUE #10783: [pickers] Digital Clock with DST time zone generates inaccurate `timeOptions` for the "fall back" and "spring forward" days
### Steps to reproduce

Link to live example: https://codesandbox.io/s/pensive-boyd-n6kq3f?file=/src/demo.tsx

Steps:

1. Open up "Fall-back DST Picker". Scroll through the list of available times
- Additional Note: `01:00 AM` and `01:30 PM` appear twice. This is a much smaller issue, but would be nice to address as well.

2. Open up "Spring-forward DST Picker". Scroll through the list of available times
- Additional note: `02:00 AM` & `02:30 AM` do no appear. However, one could argue that this is 100% correct. Works for our scenario, but just wanted to point it out for others to make that decision.

### Current behavior

When opening the "Fall-back" picker set to November 5th, 2023, the time options stop at 10:30 PM. `11:00 PM` & `11:30PM` are missing.

When opening the "Spring-forward" picker set to March 10th, 2024, the time options shows extra options for which are technically for the next day: `12:00 AM` & `12:30 AM` at the end of the time options list are both for November 11th.

### Expected behavior

When opening the "Fall-back" picker set to November 5th, 2023, all time steps of the day should be available to select. 
- Nice to have: don't show duplicate time steps (IE: `01:00 AM & `01:30 AM` in demo)

When opening the "Spring-forward" picker set to November 5th, 2023, only time steps for the selected day should be available.

### Context

The bug appears to be in the creation of `timeOptions`:
  * https://github.com/mui/mui-x/blob/v6.16.3/packages/x-date-pickers/src/DigitalClock/DigitalClock.tsx#L276
  * First time option is start of day 
  * From there, an option is created for each subsequent time step **given that there are 24 hours in a day.**

```typescript
  const timeOptions = React.useMemo(() => {
    const startOfDay = utils.startOfDay(valueOrReferenceDate);
    return [
      startOfDay,
      ...Array.from({ length: Math.ceil((24 * 60) / timeStep) - 1 }, (_, index) =>
        utils.addMinutes(startOfDay, timeStep * (index + 1)),
      ),
    ];
  }, [valueOrReferenceDate, timeStep, utils]);
```

Not completely related and 

### Your environment

<details>
  <summary><code>npx @mui/envinfo</code></summary>

```
Please see the codesandbox demo. Bug does not appear to be attributed to any browser.
```

</details>

**Search keywords**: DigitalClock timezone DST daylight luxon

### PR DESCRIPTION
**Pull request by LukasTy**
Fixes #10783

~Change `DigitalClock` time options building to use `startOfYear` to avoid `DST` issues.~
After some exploration, I decided to go with a solution proposed here: https://github.com/mui/mui-x/issues/10783#issuecomment-1777502611.
In this way we are not fighting to not have duplicate time options as they technically make sense, we only ensure that options for the whole day are included (not limiting options to `24 hours` * `minutesStep`)

### CODE REVIEW

File: packages/x-date-pickers/src/internals/utils/date-utils.ts [hunk 1]
```diff
@@ -124,6 +124,21 @@ export const mergeDateAndTime = <TDate>(
   return mergedDate;
 };
 
+export const isEqualTime = <TDate>(
```
Comment (flaviendelangle) [inline packages/x-date-pickers/src/internals/utils/date-utils.ts:127]: I am afraid this is risky if we support ms one day.
Can the ms be different? If not we could check them as well

**Review by flaviendelangle — APPROVED** [review]

**Comment by flaviendelangle** [comment]
@LukasTy what is the current status of this PR?

**Comment by LukasTy** [comment]
> what is the current status of this PR?

I shelved it while tackling more pressing matters.
Will get back to re-testing and finalizing it this week. 😉

File: packages/x-date-pickers/src/DigitalClock/DigitalClock.tsx [hunk 2]
```diff
@@ -348,10 +349,9 @@ export const DigitalClock = React.forwardRef(function DigitalClock<TDate extends
           const formattedValue = utils.format(option, ampm ? 'fullTime12h' : 'fullTime24h');
           const tabIndex =
             focusedOptionIndex === index || (focusedOptionIndex === -1 && index === 0) ? 0 : -1;
-
           return (
             <ClockItem
-              key={formattedValue}
+              key={`${option.valueOf()}-${formattedValue}`}
```
Comment (LukasTy) [inline packages/x-date-pickers/src/DigitalClock/DigitalClock.tsx:354]: Only using `option.valueOf()` fails on `AdapterDayjs` as it handles DST in a strange manner—they don't have duplicate options for the same hour, they format the duplicate offset time differently, but that's why we need to include the `formattedValue` in the key to ensure uniqueness. 🙈 
Screenshot 2024-10-24 at 12 01 22

**Comment by LukasTy** [comment]
I have updated the PR/solution and description with some extra notes.

File: packages/x-date-pickers/src/DigitalClock/DigitalClock.tsx (hunk 2 above)
Comment (flaviendelangle) [reply packages/x-date-pickers/src/DigitalClock/DigitalClock.tsx:354]: :thinking: That is weird indeed

We could also format with the day I guess, but it's probably not worth the computation time that is higher than `.valueOf` +a formatted value that we are also using elsewhere.

**Review by flaviendelangle — APPROVED** [review]
The fix make sense to me

Thanks for taking care of it :pray:

**Pull request merged**